### PR TITLE
Update of data-storage, and mark I of Tom's PID parameters

### PR DESCRIPTION
**Pull request by mschottdorf**
1. Adds derived values to the hdf5 files
2. Slight changes to the UI, for ease of use on the PI
3. O2 sensor is wired up, 
4. glitch-catcher in place, 
5. VTE estimation from flow-out with a rank-filtered baseline estimate
6. Semi-optimal PID values (Issue #123 )

### CODE REVIEW

File: vent/controller/control_module.py [hunk 1]
```diff
@@ -804,11 +836,39 @@ def _set_HAL(self, valve_open_in, valve_open_out):
...
+        glitchcatcher = True
+
+        if not glitchcatcher:
+            self._DATA_PRESSURE = self.HAL.pressure
+            self._DATA_Qout     = self.HAL.flow_ex
+            self._DATA_OXYGEN   = self.HAL.oxygen
+        
+        else:
+            pp = self.HAL.pressure
+            # if np.abs( pp  - self._DATA_PRESSURE ) < 5: # This is not a glitch, save it
```
Comment (mschottdorf) [inline vent/controller/control_module.py:850]: @zajdel trieed to catch the glitches here with these "if". Do you have a better idea?

File: sandbox/.ipynb_checkpoints/HOWTO_PlotData-checkpoint.ipynb [hunk 2]
```diff
@@ -39,8 +56,8 @@
     "    ls_ct = dff['control_data']\n",
     "    data = np.zeros((len(ls_wv),2))\n",
     "    for idx in range(len(ls_wv)):\n",
-    "        data[idx,0] = ls_wv[idx][4]  # time\n",
-    "        data[idx,1] = ls_wv[idx][3]  # pressure\n",
+    "        data[idx,0] = ls_wv[idx][6]  # time\n",
+    "        data[idx,1] = ls_wv[idx][5]  # pressure\n",
```
Comment (zajdel) [inline sandbox/.ipynb_checkpoints/HOWTO_PlotData-checkpoint.ipynb:60]: This was to reflect the changes to the logging.

File: sandbox/start_UI_pi.py [hunk 3]
```diff
@@ -5,10 +5,33 @@
...
+        app, gui = launch_gui(coordinator)
+        sys.exit(app.exec_())
+
+    except:
+        print("...ending program & closing valves")
+        
+        if not sim_mode:
+            time.sleep(0.01)
+            HAL = io.Hal( config_file = 'vent/io/config/devices.ini')
+            for i in range(10):
```
Comment (zajdel) [inline sandbox/start_UI_pi.py:29]: This termination works really well now.

File: vent/io/config/devices.ini [hunk 4]
```diff
@@ -42,6 +42,16 @@ offset_voltage    = 2.2
 output_span       = 2.0
 conversion_factor = 70.3
 
+[oxygen_sensor]
+# Note: calibration must convert to units percent O2
```
Comment (zajdel) [inline vent/io/config/devices.ini:46]: Will have to table this until we get an oxygen tank in the lab.

File: sandbox/.ipynb_checkpoints/HOWTO_PlotData-checkpoint.ipynb (hunk 2 above)
Comment (mschottdorf) [reply sandbox/.ipynb_checkpoints/HOWTO_PlotData-checkpoint.ipynb:60]: Oh, these are dictionaries; can just be done with `data[:,0] = ls_wv['time']`etc. without the loop. Should have updates this as welel

File: vent/io/config/devices.ini (hunk 4 above)
Comment (mschottdorf) [reply vent/io/config/devices.ini:46]: I guess this isn't critical for now

File: vent/controller/control_module.py [hunk 5]
```diff
@@ -774,23 +801,28 @@ def __init__(self, pid_control = False, save_logs = True, flush_every = 10, conf
         ControlModuleBase.__init__(self, pid_control, save_logs, flush_every)
         self.HAL = io.Hal(config_file)
         self._sensor_to_COPY()
-    
+
+    def __del__(self):
```
Comment (sneakers-the-rat) [inline vent/controller/control_module.py:805]: i think the order should probably be reversed for code longevity -- assume the base `__del__` method does irreparable cleanup to the object, so any derived ops should come before that.

File: vent/gui/main.py [hunk 6]
```diff
@@ -268,7 +268,7 @@ def init_ui(self):
     def init_ui_status_bar(self):
         ############
         # Status Bar
-        status_box = QtWidgets.QGroupBox('System Status')
+        status_box = QtWidgets.QGroupBox('')
```
Comment (sneakers-the-rat) [inline vent/gui/main.py:271]: these are gonna change a lot with my PR so fine to pull for now.

File: vent/gui/styles.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 CONTROL_TEXT = BACKGROUND_COLOR
 HANDLE_HEIGHT = 10
 SLIDER_WIDTH = 80
-SLIDER_HEIGHT = 50
+SLIDER_HEIGHT = 20 #50
```
Comment (sneakers-the-rat) [inline vent/gui/styles.py:21]: same thing here. sorry if these dont' work as expected. the plan is to make these values adaptive to screen size so also fine to commit for now.

File: vent/controller/control_module.py (hunk 1 above)
Comment (mschottdorf) [reply vent/controller/control_module.py:855]: added to #115 as a glitch. Guess we have to wait for now for the revised HAL

File: vent/common/loggers.py [hunk 8]
```diff
@@ -120,44 +122,62 @@ def log_exception(e, tb):
...
-    Structure for the hdf5-table to store control commands
+    Structure for the hdf5-table to store control commands. Appended whenever a control command is received.
     """
     name      = pytb.StringCol(16)   # Control setting name
     value     = pytb.Float64Col()    # double (double-precision)
     min_value = pytb.Float64Col()    # double (double-precision)
     max_value = pytb.Float64Col()    # double (double-precision)
     timestamp = pytb.Float64Col()    # double (double-precision)
 
+class DerivedSample(pytb.IsDescription):
```
[the remark this answers is not in the document]
Comment (mschottdorf) [reply vent/common/loggers.py]: I have renamed them to `ContinuousData` and `CycleData`

File: vent/common/values.py [hunk 9]
```diff
@@ -15,6 +15,25 @@ class ValueName(Enum):
     FIO2 = auto()
     VTE = auto()
     PRESSURE = auto()
+    FLOWOUT = auto()
+
+controllable_values = {
```
[the remark this answers is not in the document]
Comment (mschottdorf) [reply vent/common/values.py]: wow, actually, that wasn't even supposed to be there; good catch! This snipped of dead code must have sneaked itself back in when merged the code from the PI.

File: vent/common/loggers.py [hunk 10]
```diff
@@ -218,10 +238,21 @@ def _open_logfile(self):
         else:
             self.control_table = self.h5file.root.controls.readout
 
+        if "/derived_quantities" not in self.h5file:
```
[the remark this answers is not in the document]
Comment (mschottdorf) [reply vent/common/loggers.py:250]: done. I've also added a few more logging points (like the ran-out-of-storage-space scenario). And I've added the csv export-function that you and the matlab-export that @zajdel requested a while ago :-)

File: vent/controller/control_module.py (hunk 5 above)
Comment (mschottdorf) [reply vent/controller/control_module.py:809]: done

**Comment by mschottdorf** [comment]
General comment: (1) This is the controller that worked well on the PI. (2) I've added exports into matlab and csv. Use is pretty simple: 
```
dl = DataLogger()
dl.log2mat(filename)
dl.log2csv(filename)
```
This will produce copies of the data as .mat and .csv file in the log-folder

File: vent/common/values.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sneakers-the-rat) [reply vent/common/values.py]: yes! i thought we had gotten rid of it too, wasn't sure if it was being reintroduced when programming more in hacking mode :p

File: vent/common/loggers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sneakers-the-rat) [reply vent/common/loggers.py:250]: sweeeeet

**Comment by mschottdorf** [comment]
alright, will merge this. For the protocol: we still have to sort out the glitches

**Pull request merged**